File: app/routes/time_off.py
from datetime import datetime
from flask import Blueprint, request, jsonify
from app import db
from app.models import TimeOffRequest, Employee

bp = Blueprint("time_off", __name__, url_prefix="/api/time-off")


@bp.route("", methods=["GET"])
def list_time_off():
    query = TimeOffRequest.query

    employee_id = request.args.get("employee_id", type=int)
    if employee_id:
        query = query.filter_by(employee_id=employee_id)

    status = request.args.get("status")
    if status:
        query = query.filter_by(status=status)

    requests = query.order_by(TimeOffRequest.start_date).all()
    return jsonify([convert_to_dict(r) for r in requests]), 200


@bp.route("/<int:request_id>", methods=["GET"])
def get_time_off(request_id):
    req = TimeOffRequest.query.get_or_404(request_id)
    return jsonify(convert_to_dict(req)), 200


@bp.route("", methods=["POST"])
def create_time_off():
    data = request.get_json()

    required = ["employee_id", "start_date", "end_date"]
    missing = [f for f in required if f not in data]
    if missing:
        return jsonify({"error": f"Missing fields: {', '.join(missing)}"}), 400

    if not Employee.query.get(data["employee_id"]):
        return jsonify({"error": "Invalid employee_id"}), 400

    start_date = _parse_date(data["start_date"])
    end_date = _parse_date(data["end_date"])

    if start_date > end_date:
        return jsonify({"error": "start_date must be before or equal to end_date"}), 400

    req = TimeOffRequest(
        employee_id=data["employee_id"],
        start_date=start_date,
        end_date=end_date,
        reason=data.get("reason"),
        status="pending",
    )
    db.session.add(req)
    db.session.commit()

    return jsonify(convert_to_dict(req)), 201


@bp.route("/<int:request_id>/approve", methods=["PATCH"])
def approve_time_off(request_id):
    req = TimeOffRequest.query.get_or_404(request_id)

    if req.status != "pending":
        return jsonify({"error": f"Request is already {req.status}"}), 409

    req.status = "approved"
    db.session.commit()
    return jsonify(convert_to_dict(req)), 200


@bp.route("/<int:request_id>/reject", methods=["PATCH"])
def reject_time_off(request_id):
    req = TimeOffRequest.query.get_or_404(request_id)

    if req.status != "pending":
        return jsonify({"error": f"Request is already {req.status}"}), 409

    req.status = "rejected"
    db.session.commit()
    return jsonify(convert_to_dict(req)), 200


@bp.route("/<int:request_id>", methods=["DELETE"])
def delete_time_off(request_id):
    req = TimeOffRequest.query.get_or_404(request_id)
    db.session.delete(req)
    db.session.commit()
    return "", 204


def _parse_date(value):
    return datetime.strptime(value, "%Y-%m-%d").date()


def convert_to_dict(req):
    return {
        "id": req.id,
        "employee_id": req.employee_id,
        "employee_name": req.employee.full_name,
        "start_date": req.start_date.isoformat(),
        "end_date": req.end_date.isoformat(),
        "reason": req.reason,
        "status": req.status,
    }

File: app/static/js/time_off.js
document.addEventListener("DOMContentLoaded", async () => {
    await loadEmployeesDropdown();
    await loadRequests();

    document.getElementById("new-request-btn").addEventListener("click", openModal);
    document.getElementById("cancel-request").addEventListener("click", closeModal);
    document.getElementById("request-form").addEventListener("submit", handleSubmit);
    document.getElementById("status-filter").addEventListener("change", loadRequests);
});

async function loadEmployeesDropdown() {
    const employees = await api.get("/employees");
    const select = document.getElementById("employee-select");
    select.innerHTML = employees
        .map((e) => `<option value="${e.id}">${e.full_name}</option>`)
        .join("");
}

async function loadRequests() {
    const status = document.getElementById("status-filter").value;
    const query = status ? `?status=${status}` : "";
    const requests = await api.get(`/time-off${query}`);
    renderTable(requests);
}

function renderTable(requests) {
    const tbody = document.getElementById("time-off-body");
    tbody.innerHTML = "";

    for (const req of requests) {
        const row = document.createElement("tr");
        row.innerHTML = `
            <td>${req.employee_name}</td>
            <td>${req.start_date}</td>
            <td>${req.end_date}</td>
            <td>${req.reason || "—"}</td>
            <td><span class="status-badge status-${req.status}">${statusLabel(req.status)}</span></td>
            <td>${renderActions(req)}</td>
        `;
        tbody.appendChild(row);
    }

    document.querySelectorAll(".approve-btn").forEach((btn) => {
        btn.addEventListener("click", () => handleDecision(btn.dataset.id, "approve"));
    });
    document.querySelectorAll(".reject-btn").forEach((btn) => {
        btn.addEventListener("click", () => handleDecision(btn.dataset.id, "reject"));
    });
    document.querySelectorAll(".delete-btn").forEach((btn) => {
        btn.addEventListener("click", () => handleDelete(btn.dataset.id));
    });
}

function statusLabel(status) {
    const labels = { pending: "Εκκρεμεί", approved: "Εγκεκριμένη", rejected: "Απορρίφθηκε" };
    return labels[status] || status;
}

function renderActions(req) {
    if (req.status === "pending") {
        return `
            <button class="approve-btn" data-id="${req.id}">Έγκριση</button>
            <button class="reject-btn" data-id="${req.id}">Απόρριψη</button>
        `;
    }
    return `<button class="delete-btn" data-id="${req.id}">Διαγραφή</button>`;
}

function openModal() {
    document.getElementById("request-modal").classList.remove("hidden");
}

function closeModal() {
    document.getElementById("request-modal").classList.add("hidden");
    document.getElementById("request-form").reset();
}

async function handleSubmit(event) {
    event.preventDefault();

    const startDate = document.getElementById("start-date").value;
    const endDate = document.getElementById("end-date").value;

    const payload = {
        employee_id: Number(document.getElementById("employee-select").value),
        start_date: startDate,
        end_date: endDate,
        reason: document.getElementById("reason").value || null,
    };

    try {
        await api.post("/time-off", payload);
        closeModal();
        await loadRequests();
    } catch (err) {
        alert(err.message);
    }
}

async function handleDecision(id, decision) {
    try {
        await api.patch(`/time-off/${id}/${decision}`, {});
        await loadRequests();
    } catch (err) {
        alert(err.message);
    }
}

async function handleDelete(id) {
    if (!confirm("Διαγραφή αυτού του αιτήματος;")) return;

    try {
        await api.delete(`/time-off/${id}`);
        await loadRequests();
    } catch (err) {
        alert(err.message);
    }
}